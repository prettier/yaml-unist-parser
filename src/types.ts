import type * as YAML from "yaml";

export type ParsedCST = ReturnType<typeof YAML.parseCST>;

export interface Node {
  type: string;
  position: Position;
  /** @internal non-enumerable */
  _parent?: YamlUnistNode | null;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Range {
  origStart: number;
  origEnd: number;
}

export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Parent extends Node {
  children: Node[];
}

export interface Literal extends Node {
  value: string;
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor: null | Anchor;
  tag: null | Tag;
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
}

export interface CommentAttachable
  extends LeadingCommentAttachable,
    TrailingCommentAttachable {}

export interface LeadingCommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
}

export interface TrailingCommentAttachable {
  /** comment on the same line of the node */
  trailingComment: null | Comment;
}

export interface EndCommentAttachable {
  /** comments after the node with greater column */
  endComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | Tag
  | Anchor
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowMappingItem
  | FlowSequence
  | FlowSequenceItem;

export type ContentNode = Extract<YamlUnistNode, Content>;

// -----------------------------------------------------------------------------

export interface Comment extends Literal {
  type: "comment";
}

export interface Anchor extends Literal {
  type: "anchor";
}

export interface Tag extends Literal {
  type: "tag";
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent, TrailingCommentAttachable {
  type: "document";
  children: [DocumentHead, DocumentBody];
}

export interface DocumentHead
  extends Parent,
    EndCommentAttachable,
    TrailingCommentAttachable {
  type: "documentHead";
  children: Directive[];
}

export interface DocumentBody extends Parent, EndCommentAttachable {
  type: "documentBody";
  children: [] | [ContentNode];
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Literal, Content, CommentAttachable {
  type: "alias";
}

export interface BlockValue extends Literal, Content, LeadingCommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent: null | number;
  /** comment between indicator and the value */
  indicatorComment: null | Comment;
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Literal, Content, CommentAttachable {
  type: "plain";
}

export interface QuoteValue extends Literal, Content, CommentAttachable {}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, LeadingCommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItemBase extends Parent, LeadingCommentAttachable {
  /** key-value pair */
  children: [MappingKey, MappingValue];
}

export interface MappingItem extends MappingItemBase {
  type: "mappingItem";
}

export interface MappingKey
  extends Parent,
    TrailingCommentAttachable,
    EndCommentAttachable {
  type: "mappingKey";
  children: [] | [ContentNode];
}

export interface MappingValue
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingValue";
  children: [] | [ContentNode];
}

export interface Sequence
  extends Parent,
    Content,
    LeadingCommentAttachable,
    EndCommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItemBase extends Parent {
  children: [] | [ContentNode];
}

export interface SequenceItem
  extends SequenceItemBase,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequenceItem";
}

export interface FlowCollection
  extends Parent,
    Content,
    CommentAttachable,
    EndCommentAttachable {
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: FlowMappingItem[];
}

export interface FlowMappingItem extends MappingItemBase {
  type: "flowMappingItem";
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowSequenceItem extends SequenceItemBase {
  type: "flowSequenceItem";
}
