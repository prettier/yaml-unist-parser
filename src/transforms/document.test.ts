import { testCases, testSyntaxError } from "../helpers.js";
import { Root } from "../types.js";

const selectors = [
  getDocument(0),
  getDocumentHead(0),
  getDocumentBody(0),
  getDocument(1),
  getDocumentHead(1),
  getDocumentBody(1),
];

testSyntaxError(`- 123\na:\n`);

testCases([
  ["\n123\n\n---\n\n456\n\n", selectors],
  ["\n123\n\n...\n\n456\n\n", selectors],
  ["\n%AAA\n---\n123\n\n...\n\n456\n\n", selectors],
  ["\n%AAA\n---\n123\n\n---\n\n456\n\n", selectors],
  ["\n%AAA\n---\n123\n\n...\n\n---\n\n456\n\n", selectors],
  ["\n%AAA\n---\n123\n\n...\n\n%BBB\n---\n\n456\n\n", selectors],
  ["- AAA\n# comment\n---\n- BBB", selectors],
  ["---\nhello\n... #documentEndComment\n", getDocument(0)],
  ['&123 123 "123"\n\n... #123\n #\n\n123\n\n\n ', selectors],
  ["...\n\n#\n\n", selectors],
  ["#123\n#456\n---", getDocumentHead(0)],
  ["123\n--- #666\n456", root => root],
  ["123\n...\n456", [getDocument(0), getDocumentBody(0)]],
]);

function getDocument(documentIndex: number) {
  return (root: Root) => root.children[documentIndex];
}

function getDocumentHead(documentIndex: number) {
  return (root: Root) => getDocument(documentIndex)(root).children[0];
}

function getDocumentBody(documentIndex: number) {
  return (root: Root) => getDocument(documentIndex)(root).children[1];
}
