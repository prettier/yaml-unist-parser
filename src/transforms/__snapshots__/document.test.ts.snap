// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`"#123\\n#456\\n---" 1`] = `
documentHead (1:1 ~ 3:4)
1 | #123¶
  | ^^^^^
2 | #456¶
  | ^^^^^
3 | ---¶
  | ^^^
<documentHead>
  <endComment value="123">
  <endComment value="456">
</documentHead>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 1`] = `
document (1:1 ~ 3:4)
1 | &123·123·"123"¶
  | ^^^^^^^^^^^^^^^
2 | ¶
  | ^
3 | ...·#123¶
  | ^^^
4 | ·#¶
5 | ¶
6 | 123¶
7 | ¶
8 | ¶
9 | ·¶
<document>
  <trailingComment value="123">
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="123 \\"123\\"">
      <anchor value="123" />
    </plain>
  </documentBody>
</document>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 2`] = `
documentHead (1:1 ~ 1:1)
1 | &123·123·"123"¶
  | ~
2 | ¶
3 | ...·#123¶
4 | ·#¶
5 | ¶
6 | 123¶
7 | ¶
8 | ¶
9 | ·¶
<documentHead>

</documentHead>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 3`] = `
documentBody (1:1 ~ 2:1)
1 | &123·123·"123"¶
  | ^^^^^^^^^^^^^^^
2 | ¶
  | ^
3 | ...·#123¶
4 | ·#¶
5 | ¶
6 | 123¶
7 | ¶
8 | ¶
9 | ·¶
<documentBody>
  <plain value="123 \\"123\\"">
    <anchor value="123" />
  </plain>
</documentBody>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 4`] = `
document (4:2 ~ 9:2)
1 | &123·123·"123"¶
2 | ¶
3 | ...·#123¶
4 | ·#¶
  |  ^^
5 | ¶
  | ^
6 | 123¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
9 | ·¶
  | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="123">
      <leadingComment value="">
    </plain>
  </documentBody>
</document>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 5`] = `
documentHead (4:2 ~ 4:2)
1 | &123·123·"123"¶
2 | ¶
3 | ...·#123¶
4 | ·#¶
  |  ~
5 | ¶
6 | 123¶
7 | ¶
8 | ¶
9 | ·¶
<documentHead>

</documentHead>
`;

exports[`"&123 123 \\"123\\"\\n\\n... #123\\n #\\n\\n123\\n\\n\\n " 6`] = `
documentBody (4:2 ~ 9:2)
1 | &123·123·"123"¶
2 | ¶
3 | ...·#123¶
4 | ·#¶
  |  ^^
5 | ¶
  | ^
6 | 123¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
9 | ·¶
  | ^
<documentBody>
  <plain value="123">
    <leadingComment value="">
  </plain>
</documentBody>
`;

exports[`"...\\n\\n#\\n\\n" 1`] = `
document (1:1 ~ 1:4)
1 | ...¶
  | ^^^
2 | ¶
3 | #¶
4 | ¶
5 | ¶
<document>
  <documentHead>

  </documentHead>
  <documentBody>

  </documentBody>
</document>
`;

exports[`"...\\n\\n#\\n\\n" 2`] = `
documentHead (1:1 ~ 1:1)
1 | ...¶
  | ~
2 | ¶
3 | #¶
4 | ¶
5 | ¶
<documentHead>

</documentHead>
`;

exports[`"...\\n\\n#\\n\\n" 3`] = `
documentBody (1:1 ~ 1:1)
1 | ...¶
  | ~
2 | ¶
3 | #¶
4 | ¶
5 | ¶
<documentBody>

</documentBody>
`;

exports[`"...\\n\\n#\\n\\n" 4`] = `
document (3:1 ~ 5:1)
1 | ...¶
2 | ¶
3 | #¶
  | ^^
4 | ¶
  | ^
5 | ¶
  | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <endComment value="">
  </documentBody>
</document>
`;

exports[`"...\\n\\n#\\n\\n" 5`] = `
documentHead (3:1 ~ 3:1)
1 | ...¶
2 | ¶
3 | #¶
  | ~
4 | ¶
5 | ¶
<documentHead>

</documentHead>
`;

exports[`"...\\n\\n#\\n\\n" 6`] = `
documentBody (3:1 ~ 5:1)
1 | ...¶
2 | ¶
3 | #¶
  | ^^
4 | ¶
  | ^
5 | ¶
  | ^
<documentBody>
  <endComment value="">
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 6:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^^
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
   | ^^^
 7 | ¶
 8 | %BBB¶
 9 | ---¶
10 | ¶
11 | 456¶
12 | ¶
13 | ¶
<document>
  <documentHead>
    <directive name="AAA" parameters=[] />
  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 3:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | %BBB¶
 9 | ---¶
10 | ¶
11 | 456¶
12 | ¶
13 | ¶
<documentHead>
  <directive name="AAA" parameters=[] />
</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 3`] = `
documentBody (4:1 ~ 5:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
 7 | ¶
 8 | %BBB¶
 9 | ---¶
10 | ¶
11 | 456¶
12 | ¶
13 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 4`] = `
document (8:1 ~ 13:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | %BBB¶
   | ^^^^^
 9 | ---¶
   | ^^^^
10 | ¶
   | ^
11 | 456¶
   | ^^^^
12 | ¶
   | ^
13 | ¶
   | ^
<document>
  <documentHead>
    <directive name="BBB" parameters=[] />
  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 5`] = `
documentHead (8:1 ~ 9:4)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | %BBB¶
   | ^^^^^
 9 | ---¶
   | ^^^
10 | ¶
11 | 456¶
12 | ¶
13 | ¶
<documentHead>
  <directive name="BBB" parameters=[] />
</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n%BBB\\n---\\n\\n456\\n\\n" 6`] = `
documentBody (11:1 ~ 13:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | %BBB¶
 9 | ---¶
10 | ¶
11 | 456¶
   | ^^^^
12 | ¶
   | ^
13 | ¶
   | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 6:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^^
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
   | ^^^
 7 | ¶
 8 | ---¶
 9 | ¶
10 | 456¶
11 | ¶
12 | ¶
<document>
  <documentHead>
    <directive name="AAA" parameters=[] />
  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 3:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | ---¶
 9 | ¶
10 | 456¶
11 | ¶
12 | ¶
<documentHead>
  <directive name="AAA" parameters=[] />
</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 3`] = `
documentBody (4:1 ~ 5:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
 7 | ¶
 8 | ---¶
 9 | ¶
10 | 456¶
11 | ¶
12 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 4`] = `
document (8:1 ~ 12:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | ---¶
   | ^^^^
 9 | ¶
   | ^
10 | 456¶
   | ^^^^
11 | ¶
   | ^
12 | ¶
   | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 5`] = `
documentHead (8:1 ~ 8:4)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | ---¶
   | ^^^
 9 | ¶
10 | 456¶
11 | ¶
12 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n---\\n\\n456\\n\\n" 6`] = `
documentBody (10:1 ~ 12:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | ---¶
 9 | ¶
10 | 456¶
   | ^^^^
11 | ¶
   | ^
12 | ¶
   | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 6:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^^
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
   | ^^^
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<document>
  <documentHead>
    <directive name="AAA" parameters=[] />
  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 3:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<documentHead>
  <directive name="AAA" parameters=[] />
</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 3`] = `
documentBody (4:1 ~ 5:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ...¶
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 4`] = `
document (8:1 ~ 10:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | 456¶
   | ^^^^
 9 | ¶
   | ^
10 | ¶
   | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 5`] = `
documentHead (8:1 ~ 8:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | 456¶
   | ~
 9 | ¶
10 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n...\\n\\n456\\n\\n" 6`] = `
documentBody (8:1 ~ 10:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ...¶
 7 | ¶
 8 | 456¶
   | ^^^^
 9 | ¶
   | ^
10 | ¶
   | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 6:1)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^^
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ---¶
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<document>
  <documentHead>
    <directive name="AAA" parameters=[] />
  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 3:4)
 1 | ¶
 2 | %AAA¶
   | ^^^^^
 3 | ---¶
   | ^^^
 4 | 123¶
 5 | ¶
 6 | ---¶
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<documentHead>
  <directive name="AAA" parameters=[] />
</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 3`] = `
documentBody (4:1 ~ 6:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
   | ^^^^
 5 | ¶
   | ^
 6 | ---¶
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 4`] = `
document (6:1 ~ 10:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ---¶
   | ^^^^
 7 | ¶
   | ^
 8 | 456¶
   | ^^^^
 9 | ¶
   | ^
10 | ¶
   | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 5`] = `
documentHead (6:1 ~ 6:4)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ---¶
   | ^^^
 7 | ¶
 8 | 456¶
 9 | ¶
10 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n%AAA\\n---\\n123\\n\\n---\\n\\n456\\n\\n" 6`] = `
documentBody (8:1 ~ 10:1)
 1 | ¶
 2 | %AAA¶
 3 | ---¶
 4 | 123¶
 5 | ¶
 6 | ---¶
 7 | ¶
 8 | 456¶
   | ^^^^
 9 | ¶
   | ^
10 | ¶
   | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 4:4)
1 | ¶
2 | 123¶
  | ^^^^
3 | ¶
  | ^
4 | ...¶
  | ^^^
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 2:1)
1 | ¶
2 | 123¶
  | ~
3 | ¶
4 | ...¶
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 3`] = `
documentBody (2:1 ~ 3:1)
1 | ¶
2 | 123¶
  | ^^^^
3 | ¶
  | ^
4 | ...¶
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 4`] = `
document (6:1 ~ 8:1)
1 | ¶
2 | 123¶
3 | ¶
4 | ...¶
5 | ¶
6 | 456¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 5`] = `
documentHead (6:1 ~ 6:1)
1 | ¶
2 | 123¶
3 | ¶
4 | ...¶
5 | ¶
6 | 456¶
  | ~
7 | ¶
8 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n123\\n\\n...\\n\\n456\\n\\n" 6`] = `
documentBody (6:1 ~ 8:1)
1 | ¶
2 | 123¶
3 | ¶
4 | ...¶
5 | ¶
6 | 456¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 1`] = `
document (2:1 ~ 4:1)
1 | ¶
2 | 123¶
  | ^^^^
3 | ¶
  | ^
4 | ---¶
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 2`] = `
documentHead (2:1 ~ 2:1)
1 | ¶
2 | 123¶
  | ~
3 | ¶
4 | ---¶
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 3`] = `
documentBody (2:1 ~ 4:1)
1 | ¶
2 | 123¶
  | ^^^^
3 | ¶
  | ^
4 | ---¶
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 4`] = `
document (4:1 ~ 8:1)
1 | ¶
2 | 123¶
3 | ¶
4 | ---¶
  | ^^^^
5 | ¶
  | ^
6 | 456¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="456" />
  </documentBody>
</document>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 5`] = `
documentHead (4:1 ~ 4:4)
1 | ¶
2 | 123¶
3 | ¶
4 | ---¶
  | ^^^
5 | ¶
6 | 456¶
7 | ¶
8 | ¶
<documentHead>

</documentHead>
`;

exports[`"\\n123\\n\\n---\\n\\n456\\n\\n" 6`] = `
documentBody (6:1 ~ 8:1)
1 | ¶
2 | 123¶
3 | ¶
4 | ---¶
5 | ¶
6 | 456¶
  | ^^^^
7 | ¶
  | ^
8 | ¶
  | ^
<documentBody>
  <plain value="456" />
</documentBody>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 1`] = `
document (1:1 ~ 3:1)
1 | -·AAA¶
  | ^^^^^^
2 | #·comment¶
  | ^^^^^^^^^^
3 | ---¶
4 | -·BBB¶
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <endComment value=" comment">
    <sequence>
      <sequenceItem>
        <plain value="AAA" />
      </sequenceItem>
    </sequence>
  </documentBody>
</document>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 2`] = `
documentHead (1:1 ~ 1:1)
1 | -·AAA¶
  | ~
2 | #·comment¶
3 | ---¶
4 | -·BBB¶
<documentHead>

</documentHead>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 3`] = `
documentBody (1:1 ~ 3:1)
1 | -·AAA¶
  | ^^^^^^
2 | #·comment¶
  | ^^^^^^^^^^
3 | ---¶
4 | -·BBB¶
<documentBody>
  <endComment value=" comment">
  <sequence>
    <sequenceItem>
      <plain value="AAA" />
    </sequenceItem>
  </sequence>
</documentBody>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 4`] = `
document (3:1 ~ 4:6)
1 | -·AAA¶
2 | #·comment¶
3 | ---¶
  | ^^^^
4 | -·BBB¶
  | ^^^^^
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <sequence>
      <sequenceItem>
        <plain value="BBB" />
      </sequenceItem>
    </sequence>
  </documentBody>
</document>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 5`] = `
documentHead (3:1 ~ 3:4)
1 | -·AAA¶
2 | #·comment¶
3 | ---¶
  | ^^^
4 | -·BBB¶
<documentHead>

</documentHead>
`;

exports[`"- AAA\\n# comment\\n---\\n- BBB" 6`] = `
documentBody (4:1 ~ 4:6)
1 | -·AAA¶
2 | #·comment¶
3 | ---¶
4 | -·BBB¶
  | ^^^^^
<documentBody>
  <sequence>
    <sequenceItem>
      <plain value="BBB" />
    </sequenceItem>
  </sequence>
</documentBody>
`;

exports[`"---\\nhello\\n... #documentEndComment\\n" 1`] = `
document (1:1 ~ 3:4)
1 | ---¶
  | ^^^^
2 | hello¶
  | ^^^^^^
3 | ...·#documentEndComment¶
  | ^^^
4 | ¶
<document>
  <trailingComment value="documentEndComment">
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="hello" />
  </documentBody>
</document>
`;

exports[`"123\\n...\\n456" 1`] = `
document (1:1 ~ 2:4)
1 | 123¶
  | ^^^^
2 | ...¶
  | ^^^
3 | 456¶
<document>
  <documentHead>

  </documentHead>
  <documentBody>
    <plain value="123" />
  </documentBody>
</document>
`;

exports[`"123\\n...\\n456" 2`] = `
documentBody (1:1 ~ 1:4)
1 | 123¶
  | ^^^
2 | ...¶
3 | 456¶
<documentBody>
  <plain value="123" />
</documentBody>
`;

exports[`"123\\n--- #666\\n456" 1`] = `
root (1:1 ~ 3:4)
1 | 123¶
  | ^^^^
2 | ---·#666¶
  | ^^^^^^^^^
3 | 456¶
  | ^^^
<root>
  <document>
    <documentHead>

    </documentHead>
    <documentBody>
      <plain value="123" />
    </documentBody>
  </document>
  <document>
    <documentHead>
      <trailingComment value="666">
    </documentHead>
    <documentBody>
      <plain value="456" />
    </documentBody>
  </document>
</root>
`;

exports[`Document contains trailing content not separated by a ... or --- line 1`] = `
"Document contains trailing content not separated by a ... or --- line
1 | -·123¶
2 | a:¶
  | ^^^
3 | ¶
  | ^"
`;
